// @nebular/theme/esm2020/components/select-with-autocomplete/select-with-autocomplete.module.mjs
import { NgModule } from '@angular/core';
import { NbOverlayModule } from '../cdk/overlay/overlay.module';
import { NbSharedModule } from '../shared/shared.module';
import { NbInputModule } from '../input/input.module';
import { NbCardModule } from '../card/card.module';
import { NbButtonModule } from '../button/button.module';
import { NbSelectWithAutocompleteComponent } from './select-with-autocomplete.component';
import { NbOptionModule } from '../option/option-list.module';
import { NbSelectModule } from '../select/select.module';
import { NbIconModule } from '../icon/icon.module';
import { NbFormFieldModule } from '../form-field/form-field.module';
import * as i0 from "@angular/core";
const NB_SELECT_COMPONENTS = [NbSelectWithAutocompleteComponent];
export class NbSelectWithAutocompleteModule {
}
NbSelectWithAutocompleteModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbSelectWithAutocompleteModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NbSelectWithAutocompleteModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.4", ngImport: i0, type: NbSelectWithAutocompleteModule, declarations: [NbSelectWithAutocompleteComponent], imports: [NbSharedModule,
        NbOverlayModule,
        NbButtonModule,
        NbInputModule,
        NbCardModule,
        NbIconModule,
        NbOptionModule,
        NbFormFieldModule,
        NbSelectModule], exports: [NbSelectWithAutocompleteComponent, NbOptionModule, NbSelectModule] });
NbSelectWithAutocompleteModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbSelectWithAutocompleteModule, imports: [NbSharedModule,
        NbOverlayModule,
        NbButtonModule,
        NbInputModule,
        NbCardModule,
        NbIconModule,
        NbOptionModule,
        NbFormFieldModule,
        NbSelectModule, NbOptionModule, NbSelectModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbSelectWithAutocompleteModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        NbSharedModule,
                        NbOverlayModule,
                        NbButtonModule,
                        NbInputModule,
                        NbCardModule,
                        NbIconModule,
                        NbOptionModule,
                        NbFormFieldModule,
                        NbSelectModule,
                    ],
                    exports: [...NB_SELECT_COMPONENTS, NbOptionModule, NbSelectModule],
                    declarations: [...NB_SELECT_COMPONENTS],
                }]
        }] });
