// @nebular/theme/esm2020/components/tree-grid/tree-grid-row.component.mjs
import { Component, HostListener, Inject, Input } from '@angular/core';
import { Subject, timer } from 'rxjs';
import { take, takeUntil } from 'rxjs/operators';
import { NbCdkFooterRow, NbCdkHeaderRow, NbCdkRow } from '../cdk/table/type-mappings';
import { NbFooterRowComponent, NbHeaderRowComponent, NbRowComponent } from '../cdk/table/row';
import { NB_TREE_GRID } from './tree-grid-injection-tokens';
import * as i0 from "@angular/core";
import * as i1 from "../cdk/table/row";
export const NB_ROW_DOUBLE_CLICK_DELAY = 200;
/**
 * Cells container. Adds the right class and role.
 */
export class NbTreeGridRowComponent extends NbRowComponent {
    toggleIfEnabledNode() {
        if (!this.clickToToggle) {
            return;
        }
        timer(NB_ROW_DOUBLE_CLICK_DELAY)
            .pipe(take(1), takeUntil(this.doubleClick$))
            .subscribe(() => this.tree.toggleRow(this));
    }
    toggleIfEnabledNodeDeep() {
        if (!this.clickToToggle) {
            return;
        }
        this.doubleClick$.next();
        this.tree.toggleRow(this, { deep: true });
    }
    constructor(tree, elementRef) {
        super();
        this.elementRef = elementRef;
        this.doubleClick$ = new Subject();
        /**
         * Time to wait for second click to expand row deeply.
         * 200ms by default.
         */
        this.doubleClickDelay = NB_ROW_DOUBLE_CLICK_DELAY;
        /**
         * Toggle row on click. Enabled by default.
         */
        this.clickToToggle = true;
        this.tree = tree;
    }
    ngOnDestroy() {
        this.doubleClick$.complete();
    }
}
NbTreeGridRowComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridRowComponent, deps: [{ token: NB_TREE_GRID }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
NbTreeGridRowComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.4", type: NbTreeGridRowComponent, selector: "tr[nbTreeGridRow]", inputs: { doubleClickDelay: "doubleClickDelay", clickToToggle: "clickToToggle" }, host: { attributes: { "role": "row" }, listeners: { "click": "toggleIfEnabledNode()", "dblclick": "toggleIfEnabledNodeDeep()" }, classAttribute: "nb-tree-grid-row" }, providers: [{ provide: NbCdkRow, useExisting: NbTreeGridRowComponent }], usesInheritance: true, ngImport: i0, template: `<ng-container nbCellOutlet></ng-container>`, isInline: true, dependencies: [{ kind: "directive", type: i1.NbCellOutletDirective, selector: "[nbCellOutlet]" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridRowComponent, decorators: [{
            type: Component,
            args: [{
                    // eslint-disable-next-line @angular-eslint/component-selector
                    selector: 'tr[nbTreeGridRow]',
                    template: `<ng-container nbCellOutlet></ng-container>`,
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'nb-tree-grid-row',
                        role: 'row',
                    },
                    providers: [{ provide: NbCdkRow, useExisting: NbTreeGridRowComponent }],
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [NB_TREE_GRID]
                }] }, { type: i0.ElementRef }]; }, propDecorators: { doubleClickDelay: [{
                type: Input
            }], clickToToggle: [{
                type: Input
            }], toggleIfEnabledNode: [{
                type: HostListener,
                args: ['click']
            }], toggleIfEnabledNodeDeep: [{
                type: HostListener,
                args: ['dblclick']
            }] } });
export class NbTreeGridHeaderRowComponent extends NbHeaderRowComponent {
}
NbTreeGridHeaderRowComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridHeaderRowComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
NbTreeGridHeaderRowComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.4", type: NbTreeGridHeaderRowComponent, selector: "tr[nbTreeGridHeaderRow]", host: { attributes: { "role": "row" }, classAttribute: "nb-tree-grid-header-row" }, providers: [{ provide: NbCdkHeaderRow, useExisting: NbTreeGridHeaderRowComponent }], usesInheritance: true, ngImport: i0, template: ` <ng-container nbCellOutlet></ng-container>`, isInline: true, dependencies: [{ kind: "directive", type: i1.NbCellOutletDirective, selector: "[nbCellOutlet]" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridHeaderRowComponent, decorators: [{
            type: Component,
            args: [{
                    // eslint-disable-next-line @angular-eslint/component-selector
                    selector: 'tr[nbTreeGridHeaderRow]',
                    template: ` <ng-container nbCellOutlet></ng-container>`,
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'nb-tree-grid-header-row',
                        role: 'row',
                    },
                    providers: [{ provide: NbCdkHeaderRow, useExisting: NbTreeGridHeaderRowComponent }],
                }]
        }] });
export class NbTreeGridFooterRowComponent extends NbFooterRowComponent {
}
NbTreeGridFooterRowComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridFooterRowComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
NbTreeGridFooterRowComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "15.2.4", type: NbTreeGridFooterRowComponent, selector: "tr[nbTreeGridFooterRow]", host: { attributes: { "role": "row" }, classAttribute: "nb-tree-grid-footer-row" }, providers: [{ provide: NbCdkFooterRow, useExisting: NbTreeGridFooterRowComponent }], usesInheritance: true, ngImport: i0, template: ` <ng-container nbCellOutlet></ng-container>`, isInline: true, dependencies: [{ kind: "directive", type: i1.NbCellOutletDirective, selector: "[nbCellOutlet]" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.4", ngImport: i0, type: NbTreeGridFooterRowComponent, decorators: [{
            type: Component,
            args: [{
                    // eslint-disable-next-line @angular-eslint/component-selector
                    selector: 'tr[nbTreeGridFooterRow]',
                    template: ` <ng-container nbCellOutlet></ng-container>`,
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'nb-tree-grid-footer-row',
                        role: 'row',
                    },
                    providers: [{ provide: NbCdkFooterRow, useExisting: NbTreeGridFooterRowComponent }],
                }]
        }] });
